dhcp-controller/controllers/dhcpserver_controller.go

```diff
@@ -0,0 +1,202 @@
+	"context"
+	"github.com/go-logr/logr"
+	log "github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	dhcpv1alpha1 "dhcp-controller/api/v1alpha1"
+)
+
+// DHCPServerReconciler reconciles a DHCPServer object
+type DHCPServerReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Log    logr.Logger
+}
+
+//+kubebuilder:rbac:groups=dhcp.plumber.k8s.pf9.io,resources=dhcpservers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=dhcp.plumber.k8s.pf9.io,resources=dhcpservers/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=dhcp.plumber.k8s.pf9.io,resources=dhcpservers/finalizers,verbs=update
+//+kubebuilder:rbac:groups=apps,resources=deployments,verbs=create;delete;deletecollection;get;list;patch;update;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the DHCPServer object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.1/pkg/reconcile
+func (r *DHCPServerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	//_ = log.FromContext(ctx)
+
+	//log := r.Log.WithValues("dhcpservers", req.NamespacedName)
+
+	var dhcpConfigReq = dhcpv1alpha1.DHCPServerList{}
+	if err := r.List(ctx, &dhcpConfigReq); err != nil {
+		log.Error(err, "unable to fetch DHCPServer")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log.Info("Reconciling")
+
+	serverList := dhcpConfigReq.Items
+	if len(serverList) > 0 {
+		for _, server := range serverList {
+			log.Infof("server %s", server.Name)
+			result, err := r.ensureServer(req, server, r.backendDeployment(server))
+			if result != nil {
+				log.Error(err, "DHCP server Not ready")
+				return *result, err
+			}
+		}
+	} else {
+		log.Info("serverList is Empty")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *DHCPServerReconciler) ensureServer(request reconcile.Request,
+	server dhcpv1alpha1.DHCPServer,
+	dep *appsv1.Deployment,
+) (*reconcile.Result, error) {
+
+	// See if deployment already exists and create if it doesn't
+	found := &appsv1.Deployment{}
+	err := r.Get(context.TODO(), types.NamespacedName{
+		Name:      dep.Name,
+		Namespace: server.Namespace,
+	}, found)
+	if err != nil && errors.IsNotFound(err) {
```

Comment:
So this just checks if we find a DHCP deployment with name matching. But what if a deployment is found but some parameters like network name, interface, or configmap changed? Don't you want to redeploy it as well?